Enhancement of Indicator Collect

kelby-shelton opened the pull request:
- Added 5 new parameters
  - `artifact_ids_include` to only search for certain artifacts for indicators
  - `indicator_types_include` to only include indicators with certain  data types (e.g. hash, domain)
  - `indicator_types_exclude` is opposite of indicator_types_include
  - `indicator_tags_include` to only include certain indicators that have the specific list of tags (e.g. malware, suspicious).
  - `indicator_tags_exclude` is opposite of indicator_tags_include

kelby-shelton committed:
Enhancement of Indicator Collect

- Added 5 new parameters
  - artifact_ids_include to only search for certain artifacts for indicators
  - indicator_types_include to only include indicators with certain  data types (e.g. hash, domain)
  - indicator_types_exclude is opposite of indicator_types_include
  - indicator_tags_include to only include certain indicators that have the specific list of tags (e.g. malware, suspicious).
  - indicator_tags_exclude is opposite of indicator_tags_include

philroyer-phantom commented on custom_functions/indicator_collect.json:
```diff
@@ -12,6 +12,43 @@
+            "contains_type": [
+                "phantom artifact id"
+            ],
+            "description": "Optional parameter to only look for indicator values that occur in the artifacts with these IDs. Must be one of: json serializable list, comma separated integers, or a single integer.",
```
Not sure about the use case for accepting a single integer artifact ID, but if we do want to do that I think we have a problem because its not working at the moment.

philroyer-phantom commented on custom_functions/indicator_collect.py:
```diff
@@ -52,37 +70,106 @@ def indicator_collect(container=None, **kwargs):
+            raise TypeError(
+                f"Invalid artifact_ids_include entered: '{artifact_ids_include}'. Must be one of: json seriablize, comma separated, or an integer."
+            )
+        elif isinstance(artifact_ids_include, list) and not check_numeric_list(artifact_ids_include):
```
If we want to accept a single integer artifact ID here, we need to check for an int and put it in a list.

kelby-shelton committed:
Fixed artifact_ids_include parameter

kelby-shelton committed:
Code rewrite

- Eliminated code reuse by making the include_exclude method more generic
- Moved data type checking to the artifact level before requesting indicator value details. This leads to a large performance gain.
- Eliminated need for a cef dictionary to be built.

kelby-shelton committed:
Rewrote code to use _filter with __in

philroyer-phantom commented on custom_functions/indicator_collect.py:
```diff
@@ -20,30 +25,38 @@ def indicator_collect(container=None, **kwargs):
+    
+    def get_indicator_json(value_list):
+        indicator_url = phantom.build_phantom_rest_url('indicator') + '?page_size=0&timerange=all'
+        hashed_list = [sha256(item.encode('utf-8')).hexdigest() for item in value_list]
```
This line is causing an encoding exception when the cef_value is actually a list of strings or an integer. I think a simple fix might be to add a check for string further down on line 113.

philroyer-phantom commented on custom_functions/indicator_collect.py:
```diff
@@ -52,40 +65,93 @@ def indicator_collect(container=None, **kwargs):
+            for cef_key in artifact['cef']:
+                cef_value = artifact['cef'][cef_key]
+                data_types = artifact['cef_types'].get(cef_key, [])
 
```
I suggest we add a check for string values only here.

`if isinstance(cef_value, str):`

philroyer-phantom commented on custom_functions/indicator_collect.py:
```diff
@@ -52,40 +65,93 @@ def indicator_collect(container=None, **kwargs):
+                    is_valid_indicator(indicator_types_exclude, data_types, check_type='exclude')
+                    and is_valid_indicator(indicator_types_include, data_types, check_type='include')
+                ):
+                    indicator_list.append(cef_value)
```
Wouldn't it make sense to deduplicate these cef values here?

`if cef_value not in indicator_list:`

kelby-shelton committed:
Changed accepted values to indicator id translation

philroyer-phantom reviewed (approved):
Looks good!
